src/messages.cpp

```diff
@@ -238,159 +241,235 @@ bool Messages::has_undisplayed_messages()
+     * '-max_padlength' the characters of space that are allocated to time units (e.g. "years", "minutes", etc.)
+     *
+     * 'bottom' works much like 'maxlength', but instead it refers to the amount of lines that the message box may hold.
+     */
+    const int maxlength = FULL_SCREEN_WIDTH - 2 - 1 - 4 - max_padlength;
     const int bottom = FULL_SCREEN_HEIGHT - 2;
     const int msg_count = size();
 
-    for (;;) {
-        werase(w);
-        draw_border(w);
-        mvwprintz(w, bottom + 1, 32, c_red, _("Press %s to return"), ctxt.get_desc("QUIT").c_str());
-        draw_scrollbar(w, offset, bottom, msg_count, 1);
+    /* Dealing With Scroll Direction
+     * =============================
+     * Much like how the sidebar can have variable scroll direction, so will the message box.
+     * To properly differentiate the two methods of displaying text, we will label them NEWEST-TOP, and OLDEST-TOP. This labeling should be self explanatory.
+     *
+     * Note that 'offset' tracks only our current position in the list; it shan't at all affect the manner in which the messages are drawn.
+     * Messages are always drawn top-to-bottom. If NEWEST-TOP is used, then the top line (line=1) corresponds to the newest message. The one further down the second-newest, etc.
+     * If the OLDEST-TOP method is used, then the top line (line=1) corresponds to the oldest message, and the bottom one to the newest.
+     * The 'for (;;)' block below is nearly completely method-agnostic, save for the `player_messages.impl_->history(i)` call.
+     *
+     * In case of NEWEST-TOP, the 'i' variable easily enough corresponds to the newest message.
+     * In case of OLDEST-TOP, the 'i' variable must be flipped- meaning the highest value of 'i' returns the result for the lowest value of 'i', etc.
+     * To achieve this, the 'flip_message' variable is set to either the value of 'msg_count', or '0'. This variable is then subtracted from 'i' in each call to player_messages.impl_->history();
+     *
+     * 'offset' refers to the corresponding message that will be displayed at the very TOP of the message box window.
+     *  NEWEST-TOP: 'offset' starts simply at '0' - the very top of the window.
+     *  OLDEST-TOP: 'offset' is set to the maximum value it could possibly be. That is- 'msg_count-bottom'. This way, the screen starts with the scrollbar all the way down.
+     * 'retrieve_history' refers to the line that should be displayed- this is either 'i' if it's NEWEST-TOP, or a flipped version of 'i' if it's OLDEST-TOP.
+     */
+    int offset = log_from_top ? 0 : ( msg_count - bottom );
+    const int flip = log_from_top ? 0 : msg_count - 1;
+    int retrieve_history;
+
+    for( ;; ) {
+        werase( w );
+        draw_border( w );
+        mvwprintz( w, bottom + 1, 32, c_red, _( "Press %s to return" ), ctxt.get_desc( "QUIT" ).c_str() );
+        draw_scrollbar( w, offset, bottom, msg_count - bottom, 1 );
```

Comment:
Please do not pass `msg_count - bottom` as the message count, since it could cause weird behavior for the scrollbar. Something like
```c++
draw_scrollbar( w, offset, bottom, msg_count, 1, 0, c_white, true );
```
is more appropriate. The last flag is specifically designed for this purpose.